### PR TITLE
Implement prepareRun method for Transform

### PR DESCRIPTION
**Pull request by anwar6953**
Motivation: Transform should be able to write to TMS once (at the beginning of a batch job).

Implement prepareRun and onRunFinish methods for Transform. The prepareRun will get called once at the beginning of the batch job and the onRunFinish will get called at the end of the batch job.
Currently, they expose a context which is the same as TransformContext, but additionally exposes access to the current namespace as well as implementing methods of `MessagingContext`.

Related: https://issues.cask.co/browse/CDAP-10974

https://builds.cask.co/browse/CDAP-RUT1234-1

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Transform.java [hunk 1]
```diff
@@ -61,7 +62,29 @@ public void destroy() {
...
+    // no-op
+  }
+
+  /**
+   * Invoked after the Batch run finishes. Used to perform any end of the run logic.
+   *
+   * @param succeeded defines the result of batch execution: true if run succeeded, false otherwise
+   * @param context transform prepare context
+   */
+  public void onRunFinish(boolean succeeded, TransformPrepareContext context) {
```
Comment (anwar6953) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Transform.java]: `TransformPrepareContext` needs to be renamed, since its naming doesn't fit the usage here (in `onRunFinish` method).

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Transform.java [hunk 2]
```diff
@@ -61,7 +62,29 @@ public void destroy() {
     //no-op
   }
 
+  /**
+   * Prepare the Batch run. Used to configure the job before starting the run.
+   *
+   * @param context transform prepare context
+   * @throws Exception if there's an error during this method invocation
+   */
+  public void prepareRun(TransformPrepareContext context) throws Exception {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Transform.java]: it seems like these methods should be a part of StageLifecycle and supported for all plugin types, not just transforms, source, and sinks.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package co.cask.cdap.etl.api;
+
+import co.cask.cdap.api.messaging.MessagingContext;
+
+/**
+ *
+ */
+public interface TransformPrepareContext extends TransformContext, MessagingContext {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java]: SubmitterContext maybe? naming is hard.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.api;
+
+import co.cask.cdap.api.messaging.MessagingContext;
+
+/**
+ *
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java]: this is in API, needs a javadoc. Should mention avoiding publishing a lot of messages to TMS, as overloading it will cause cluster instability.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public interface TransformPrepareContext extends TransformContext, MessagingContext {
+
+  /**
+   * @return The application namespace
+   */
+  String getNamespace();
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java]: seems like we should just add this to StageContext

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Transform.java (hunk 2 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/Transform.java]: Or we could put the methods in a separate SubmitterLifecycle interface or something like that, to try and make it clearer that they are called from different processes.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java (hunk 4 above)
Comment (anwar6953) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java]: I think the javadoc belongs either in MessagingContext or MessagePublisher APIs. I'll include it there.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java (hunk 3 above)
Comment (anwar6953) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/TransformPrepareContext.java]: How about just StageSubmitter?

**Comment by anwar6953** [comment]
Putting this on hold; will resume work on it as soon as I finish the implementation for https://issues.cask.co/browse/CDAP-12084.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitter.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+package co.cask.cdap.etl.api;
+
+import co.cask.cdap.api.messaging.MessagingContext;
+
+/**
+ * Similar to {@link TransformContext}, but also exposing functionality of {@link MessagingContext}.
+ *
+ * @param <T> execution context
+ */
+public interface StageSubmitter<T> extends MessagingContext {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitter.java]: This seems like an unnecessary interface. Shouldn't it be another context? Everywhere else, plugins are just given a context object.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/SubmitterLifecycle.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+@Beta
+public interface SubmitterLifecycle<T> {
+
+  /**
+   * Prepare the run. Used to configure the job before starting the run.
+   *
+   * @param context submitter context
+   * @throws Exception if there's an error during this method invocation
+   */
+  void prepareRun(StageSubmitter<T> context) throws Exception;
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/SubmitterLifecycle.java]: it seems like we should just pass the context T in here instead of a `StageSubmitter<T>`

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSource.java [hunk 8]
```diff
@@ -37,7 +38,8 @@
  */
 @Beta
 public abstract class BatchSource<KEY_IN, VAL_IN, OUT> extends BatchConfigurable<BatchSourceContext>
-  implements Transformation<KeyValue<KEY_IN, VAL_IN>, OUT>, StageLifecycle<BatchRuntimeContext> {
+  implements Transformation<KeyValue<KEY_IN, VAL_IN>, OUT>, StageLifecycle<BatchRuntimeContext> { // ,
+//  SubmitterLifecycle<BatchSourceContext> {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSource.java]: if we don't have StageSubmitter, BatchConfigurable can be changed to implement SubmitterLifecycle and you can remove this comment

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java [hunk 9]
```diff
@@ -203,15 +205,15 @@ public void initialize() throws Exception {
...
+    for (String sourceName : phase.getSources()) {
       try {
         BatchConfigurable<BatchSourceContext> batchSource = pluginInstantiator.newPluginInstance(sourceName, evaluator);
         StageSpec stageSpec = phaseSpec.getPhase().getStage(sourceName);
         MapReduceBatchContext sourceContext = new MapReduceBatchContext(context, pipelineRuntime, stageSpec,
                                                                         connectorDatasets);
         batchSource.prepareRun(sourceContext);
-        for (String inputAlias : sourceContext.getInputNames()) {
-          inputAliasToStage.put(inputAlias, sourceName);
+        for (Object inputAlias : sourceContext.getInputNames()) {
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java:214]: why this change to Object?

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceBatchContext.java [hunk 10]
```diff
@@ -163,4 +169,25 @@ private Output getOutput(Output output) {
     }
     return output;
   }
+
+  @Override
+  public MessagePublisher getMessagePublisher() {
+    // TODO: use caller?
```
Comment (albertshau) [inline cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceBatchContext.java]: should use it, it's used to unset MDC keys in the logger so system logs are not tagged with pipeline MDC keys

File: cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java (hunk 9 above)
Comment (anwar6953) [reply cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/mapreduce/ETLMapReduce.java:214]: My Intellij was complaining otherwise... I'll have to look into it.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitter.java (hunk 6 above)
Comment (anwar6953) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/StageSubmitter.java]: Yea, this was meant as a context object. Will rename it as such.

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceBatchContext.java (hunk 10 above)
Comment (anwar6953) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceBatchContext.java]: Would it help or make a difference if the implementation is just a getter?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/SubmitterLifecycle.java (hunk 7 above)
Comment (anwar6953) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/SubmitterLifecycle.java]: done

File: cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceBatchContext.java (hunk 10 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/batch/mapreduce/MapReduceBatchContext.java]: oh you're right, there's no need for that. can remove the todo then.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSource.java (hunk 8 above)
Comment (albertshau) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/co/cask/cdap/etl/api/batch/BatchSource.java]: remove this comment

**Review by albertshau — APPROVED** [review]
one comment about removing comments from the api class, otherwise lgtm. fix and merge.

**Review by albertshau — APPROVED** [review]
lgtm
